app/models/name/propagate_generic_classifications.rb

```diff
@@ -0,0 +1,130 @@
+      Name.select(:id, :synonym_id, :text_name, :classification).
+        where(rank: 0..(Name.ranks[:Genus] - 1)).
+        each do |name|
+          old_class = old_classification(name)
+          new_class = new_classification(name, accepted_names, classifications)
+          next if old_class == new_class
+
+          fixes << [name, old_class, new_class]
+        end
+      execute_propagation_fixes(fixes, dry_run)
+    end
+
+    private
+
+    def old_classification(name)
+      str = name.classification
+      str.present? && str.strip
+    end
+
+    def new_classification(name, accepted_names, classifications)
+      accepted_text_name = accepted_names[name.synonym_id] || name.text_name
+      genus = accepted_text_name.split.first
+      str = classifications[genus]
+      str.present? && str.strip
+    end
+
+    def build_accepted_names_lookup_table
+      accepted_names = {}
+      Name.select(:synonym_id, :text_name).
+        where(rank: 0..Name.ranks[:Genus], deprecated: false).
+        where.not(synonym_id: nil).
+        each do |name|
+          accepted_names[name.synonym_id] = name.text_name
+        end
+      accepted_names
+    end
+
+    def accepted_generic_classification_strings
+      classifications = {}
+      Name.select(:text_name, :classification).
+        where(rank: Name.ranks[:Genus], deprecated: false).
+        where("author NOT LIKE 'sensu lato%'").
+        each do |name|
+          next if name.classification.blank?
+
+          if classifications[name.text_name].present?
+            warn("Multiple accepted non-sensu lato genera for #{text_name}!")
+          else
+            classifications[name.text_name] = name.classification
+          end
+        end
+      classifications
+    end
+
+    def hash_of_names_with_observations
+      Hash[
+        Observation.distinct.pluck(:text_name).collect do |text_name|
+          [text_name, true]
+        end
+      ]
+    end
+
+    def execute_propagation_fixes(fixes, dry_run)
+      bundles = {}
+      used_names = hash_of_names_with_observations
+      msgs = fixes.map do |name, old_class, new_class|
+        bundles[new_class] = [] if bundles[new_class].blank?
+        bundles[new_class] << name.id
+        next unless used_names[name.text_name]
+
+        describe_propagation_fix(name, old_class, new_class)
+      end
+      msgs.reject(&:nil?) + execute_bundled_propagation_fixes(bundles, dry_run)
+    end
+
+    def execute_bundled_propagation_fixes(bundles, dry_run)
+      msgs = []
+      bundles.each do |classification, ids|
+        next if classification.blank?
+
+        msgs << "Setting classifications for #{ids.join(",")}"
+        next if dry_run
+
+        # Deliberately skip validations
+        # rubocop:disable Rails/SkipsModelValidations
+        Name.where(id: ids).update_all(classification: classification)
+        # rubocop:enable Rails/SkipsModelValidations
+      end
+      msgs
```

Comment:
```suggestion
      bundles.each_with_object([]) do |bundle, msgs|
        # bundle => [<classification string>, [array of Name ids]]
        ids = bundle[1]
        msgs << "Setting classifications for #{bundle[1].join(",")}"
        next if dry_run

        # Deliberately skip validations
        # rubocop:disable Rails/SkipsModelValidations
        Name.where(id: bundle[1]).
          update_all(classification: bundle[0])
        # rubocop:enable Rails/SkipsModelValidations
```
See above re `each_with_object`. The syntax here is awkward enough so that I added a comment.